### PR TITLE
[PM-41690] fix: CXF import crash on negative timestamps

### PR DESCRIPTION
**Pull request by crosenth**
Some credential managers (e.g., Google Password Manager) export creationAt/modifiedAt as the Windows FILETIME epoch (-11644473600) when no real date exists. The credential-exchange-format crate deserializes these fields as u64 and rejects negative values.

Sanitize the JSON payload before deserialization by clamping negative timestamp values to `null`.

## 🎟️ Tracking

https://github.com/bitwarden/android/issues/7140

https://github.com/bitwarden/android/issues/7215

https://github.com/bitwarden/android/pull/7216

https://bitwarden.atlassian.net/browse/PM-40542

## Summary

  - Sanitize negative `creationAt`/`modifiedAt` values in CXF payloads
    before deserialization, clamping them to `null`
  - Fixes import failures from Google Password Manager on devices where
    credentials have no real creation/modification date (exported as the
    Windows FILETIME epoch: -11644473600)

  ## Root cause

  The `credential-exchange-format` crate defines these fields as
  `Option<u64>`, which cannot represent negative values. The fix
  pre-processes the JSON in `parse_cxf()` before the typed
  deserialization step.

  ## Test plan

  - Added unit tests for `sanitize_timestamps()` covering negative,
    valid, and mixed timestamp values
  - Added integration test confirming `parse_cxf()` succeeds with
    negative timestamps in the payload

### CODE REVIEW

File: crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs
Comment (harr1424) [inline crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs:1]: 🎨  Thanks for adding comprehensive test coverage. As described in another comment, this PR does not resolve the crash when these negative values are nested within a `collections` array as opposed to an `items` array: 

```rust
#[test]
fn poc_collections_negative_timestamp_still_crashes() {
    let input = r#"{
        "id": "DZSXp7iBQY-Fg-OofakQtQ", "username": "u", "email": "e",
        "collections": [{
            "id": "DZSXp7iBQY-Fg-OofakQtQ",
            "creationAt": -11644473600,
            "modifiedAt": -11644473600,
            "title": "C",
            "items": []
        }],
        "items": []
    }"#;
    let result = parse_cxf(input.to_string());
    assert!(result.is_ok(), "{:?}", result.err());
}
```

🎨 If you agree with the code changes I've suggested in `import.rs` I would also suggest adding the following test coverage: 
```rust
#[test]
fn test_sanitize_valid_timestamps_unchanged_returns_borrowed() {
    let input = r#"{"id":"test","items":[{"id":"1","creationAt":1759783057,"modifiedAt":1759783057,"title":"Test","credentials":[]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(matches!(result, std::borrow::Cow::Borrowed(_)));
}

#[test]
fn test_sanitize_negative_timestamp_in_collection() {
    let input = r#"{"id":"test","items":[],"collections":[{"id":"c1","creationAt":-11644473600,"modifiedAt":-11644473600,"title":"C","items":[]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(result.contains(r#""creationAt":null"#));
    assert!(result.contains(r#""modifiedAt":null"#));
}

#[test]
fn test_sanitize_negative_timestamp_in_sub_collection() {
    let input = r#"{"id":"test","items":[],"collections":[{"id":"c1","title":"C","items":[],
        "subCollections":[{"id":"c2","creationAt":-11644473600,"title":"Sub","items":[]}]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(result.contains(r#""creationAt":null"#));
}

#[test]
fn test_parse_cxf_with_negative_timestamps_falls_back_to_current_time() {
    let input = r#"{
        "id": "DZSXp7iBQY-Fg-OofakQtQ", "username": "u", "email": "e",
        "collections": [{"id":"c1","creationAt":-11644473600,"modifiedAt":-11644473600,"title":"C","items":[]}],
        "items": [{
            "id": "9OF-QjVDQo2Wp2xWPw6ZhA",
            "creationAt": -11644473600, "modifiedAt": -11644473600,
            "title": "Test Entry",
            "credentials": [{"type":"basic-auth","username":{"id":"u","fieldType":"string","value":"testuser"},
                              "password":{"id":"p","fieldType":"concealed-string","value":"testpass"},
                              "urls":["https://example.com"]}]
        }]
    }"#;
    let ciphers = parse_cxf(input.to_string()).expect("should not error on negative timestamps");
    let cipher = ciphers.first().unwrap();
    assert!(cipher.creation_date > Utc::now() - chrono::Duration::seconds(5));
}
```

🎨 You will also want to update existing tests in this file to expect `null` instead of `0`

File: crates/bitwarden-exporters/src/cxf/import.rs [hunk 1]
```diff
@@ -29,13 +29,41 @@ use crate::{
...
+                        item[key] = serde_json::Value::Number(0.into());
+                    }
+                }
+            }
+        }
+    }
+
+    serde_json::to_string(&value).unwrap_or_else(|_| payload.to_string())
+}
+
```
Comment (harr1424) [inline crates/bitwarden-exporters/src/cxf/import.rs]:
```suggestion
pub(crate) fn sanitize_timestamps(payload: &str) -> std::borrow::Cow<'_, str> {
    let Ok(mut value) = serde_json::from_str::<serde_json::Value>(payload) else {
        return std::borrow::Cow::Borrowed(payload);
    };

    let mut modified = false;

    if let Some(items) = value.get_mut("items").and_then(|v| v.as_array_mut()) {
        for item in items {
            clamp_timestamps(item, &mut modified);
        }
    }
    if let Some(collections) = value.get_mut("collections").and_then(|v| v.as_array_mut()) {
        for collection in collections {
            clamp_collection_timestamps(collection, &mut modified);
        }
    }

    if !modified {
        return std::borrow::Cow::Borrowed(payload);
    }
    serde_json::to_string(&value)
        .map(std::borrow::Cow::Owned)
        .unwrap_or(std::borrow::Cow::Borrowed(payload))
}

fn clamp_timestamps(item: &mut serde_json::Value, modified: &mut bool) {
    for key in ["creationAt", "modifiedAt"] {
        if item.get(key).and_then(|v| v.as_i64()).is_some_and(|n| n < 0) {
            item[key] = serde_json::Value::Null;
            *modified = true;
        }
    }
}

fn clamp_collection_timestamps(collection: &mut serde_json::Value, modified: &mut bool) {
    clamp_timestamps(collection, modified);
    if let Some(subs) = collection.get_mut("subCollections").and_then(|v| v.as_array_mut()) {
        for sub in subs {
            clamp_collection_timestamps(sub, modified);
        }
    }
}

```

Comment (harr1424) [reply crates/bitwarden-exporters/src/cxf/import.rs]: 1. Clone on write paired with early return when no values are modified avoids `serde_json::to_string()` serialization at every parse 
2. Assignment of `serde_json::Value::Null` ensures CXF specs are respected, as `null` will fallback to `Utc::now()`
3. Parse and check `collections` recursively

**Review by harr1424 — CHANGES_REQUESTED** [review]
@crosenth This PR and the research you invested in it is greatly appreciated. I do disagree with the fundamental approach taken here, which diverges from [v1.0 specifications](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#entity-collection): 

>creationAt
This OPTIONAL member contains the UNIX timestamp in seconds at which this [Collection](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#cddl-type-collection) was originally created. If this member is not set, but the importing provider requires this member in their proprietary data model, the importer SHOULD use the current timestamp at the time the provider encounters this [Collection](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#cddl-type-collection).

>modifiedAt
This OPTIONAL member contains the UNIX timestamp in seconds of the last modification brought to this [Collection](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#cddl-type-collection). If this member is not set, but the importing provider requires this member in their proprietary data model, the importer SHOULD use the current timestamp at the time the provider encounters this [Collection](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#cddl-type-collection).

[The above also applies to these values nested in the `items` array](https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#entity-item). 

Additionally, the fix has only been applied to negative values in the `items` array and not `collection` which can also introduce the negative values and cause a crash.  Please see my comment in `crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs` for a POC. 

I've suggested code changes to improve the efficiency of the timestamp adjustment and ensure it results in the current timestamp. 

I've also suggested some additional test coverage.

File: crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs
Comment (crosenth) [reply crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs:1]: Thanks @harr1424  - Merged in your updates and fixed the `null` instead of `0` issue

File: crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs [hunk 2]
```diff
@@ -39,6 +39,14 @@ mod tests {
...
     }
 
+    #[test]
+    fn test_sanitize_no_modification_returns_original() {
+        let input = r#"{"id":"test","items":[{"id":"1","creationAt":1759783057,"modifiedAt":1759783057,"title":"Test","credentials":[]}]}"#;
+        let result = sanitize_timestamps(input);
+        // When no modification is needed, should return a borrowed reference to the original
+        assert_eq!(result.as_ref(), input);
+    }
+
```
Comment (harr1424) [inline crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs:50]: @crosenth This is a valuable test to add, but considering how the `sanitize_timestamps()` function was modified, I still think that adding the following tests from my original comment would be of value: 

```rust
#[test]
fn test_sanitize_valid_timestamps_unchanged_returns_borrowed() {
    let input = r#"{"id":"test","items":[{"id":"1","creationAt":1759783057,"modifiedAt":1759783057,"title":"Test","credentials":[]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(matches!(result, std::borrow::Cow::Borrowed(_)));
}

#[test]
fn test_sanitize_negative_timestamp_in_collection() {
    let input = r#"{"id":"test","items":[],"collections":[{"id":"c1","creationAt":-11644473600,"modifiedAt":-11644473600,"title":"C","items":[]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(result.contains(r#""creationAt":null"#));
    assert!(result.contains(r#""modifiedAt":null"#));
}

#[test]
fn test_sanitize_negative_timestamp_in_sub_collection() {
    let input = r#"{"id":"test","items":[],"collections":[{"id":"c1","title":"C","items":[],
        "subCollections":[{"id":"c2","creationAt":-11644473600,"title":"Sub","items":[]}]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(result.contains(r#""creationAt":null"#));
}

#[test]
fn test_parse_cxf_with_negative_timestamps_falls_back_to_current_time() {
    let input = r#"{
        "id": "DZSXp7iBQY-Fg-OofakQtQ", "username": "u", "email": "e",
        "collections": [{"id":"c1","creationAt":-11644473600,"modifiedAt":-11644473600,"title":"C","items":[]}],
        "items": [{
            "id": "9OF-QjVDQo2Wp2xWPw6ZhA",
            "creationAt": -11644473600, "modifiedAt": -11644473600,
            "title": "Test Entry",
            "credentials": [{"type":"basic-auth","username":{"id":"u","fieldType":"string","value":"testuser"},
                              "password":{"id":"p","fieldType":"concealed-string","value":"testpass"},
                              "urls":["https://example.com"]}]
        }]
    }"#;
    let ciphers = parse_cxf(input.to_string()).expect("should not error on negative timestamps");
    let cipher = ciphers.first().unwrap();
    assert!(cipher.creation_date > Utc::now() - chrono::Duration::seconds(5));
}
```

Comment (crosenth) [reply crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs:50]: Agreed thanks @harr1424:  https://github.com/bitwarden/sdk-internal/pull/1362/changes/9f9c360fb734cd56c6f8945a7766cf507045df8a

File: crates/bitwarden-exporters/src/cxf/import.rs [hunk 3]
```diff
@@ -29,13 +29,64 @@ use crate::{
...
-    let account: CxfAccount = serde_json::from_str(&payload)?;
+    let sanitized = sanitize_timestamps(&payload);
+    let account: CxfAccount = serde_json::from_str(&sanitized)?;
 
     let items: Vec<ImportingCipher> = account.items.into_iter().flat_map(parse_item).collect();
 
     Ok(items)
 }
 
+/// Replace negative `creationAt` and `modifiedAt` values with 0.
```
Comment (harr1424) [inline crates/bitwarden-exporters/src/cxf/import.rs]:
```suggestion
/// Replace negative `creationAt` and `modifiedAt` values with null, so that `Utc::now()` will be used downstream 
```

File: crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs
Comment (harr1424) [reply crates/bitwarden-exporters/src/cxf/tests/negative_timestamp_test.rs:1]: ❓ @crosenth Is there any rationale for omitting the following test coverage, which has been suggested across two review cycles? My rationale for adding such is that it guards against any future code changes here that could take ownership at every parse, with the changes you accepted to  `sanitize_timestamps()` ownership is only taken when the payload needs to be adjusted to avoid the crash. 

```rust
#[test]
fn test_sanitize_valid_timestamps_unchanged_returns_borrowed() {
    let input = r#"{"id":"test","items":[{"id":"1","creationAt":1759783057,"modifiedAt":1759783057,"title":"Test","credentials":[]}]}"#;
    let result = sanitize_timestamps(input);
    assert!(matches!(result, std::borrow::Cow::Borrowed(_)));
}
```

**Review by harr1424 — CHANGES_REQUESTED** [review]
I've left a suggestion to update a doc comment to describe the new behavior, please also update the PR description as it still describes the old behavior (assigning 0 instead of null). 

Please also be sure to run the [checks described here](https://github.com/bitwarden/sdk-internal#running-the-checks) as your PR won't be able to merge until these pass. 

Lastly I left a comment in the new test file asking if there was any reason not to add a test ensuring allocation is avoided when the payload doesn't need to be modified. If you'd prefer not to add this, it's probably OK, it would be nice to document your reasoning, though.

**Review by harr1424 — DISMISSED** [review]

**Comment by harr1424** [comment]
@crosenth please be sure to run the lint [described here](https://github.com/bitwarden/sdk-internal#running-the-checks) in order to format these changes against standards, right now the required CI check "Lint / Check Style (pull_request)" is failing.

You can run `npm run lint:fix` from the root of your cloned repository.

**Review by harr1424 — APPROVED** [review]

**Comment by harr1424** [comment]
@crosenth we are waiting on a second code-review from a colleague on my team. Everything looks good on my end and I look forward to merging your changes as soon as they are reviewed by a second pair of eyes. 

@adudek-bw can you please give this a review?

**Comment by crosenth** [comment]
> @crosenth we are waiting on a second code-review from a colleague on my team. Everything looks good on my end and I look forward to merging your changes as soon as they are reviewed by a second pair of eyes.
> 
> @adudek-bw can you please give this a review?

Thanks @harr1424 really enjoyed iterating on this with you

**Review by adudek-bw — APPROVED** [review]
